### ISSUE TITLE
Unable to handle BIGINT on 32 Bit arm

### ISSUE THREAD

**Issue by alanj853** [elixir-sqlite/exqlite#196]
Hi,

I am currently working on a custom armv7 board (similar to a raspberry pi), and am having trouble upgrading to using exqlite. I was using [esqlite](https://github.com/mmzeeman/esqlite), and everything was working fine, but due to the fact that https://github.com/elixir-sqlite/sqlitex and https://github.com/elixir-sqlite/sqlite_ecto2 are no longer being maintained and needed to use ecto 2, I am trying to switch over to the [exqlite](https://github.com/elixir-sqlite/exqlite), [ecto_sqlite3](https://github.com/elixir-sqlite/ecto_sqlite3) combination where I can also get ecto 3.

The issue in exqlite is failing to handle integers greater than 32 bit in size during a schema migration, although esqlite can handle them. For example, for the integer 20180801160101, -1250163803 is being returned when doing a multistep.

The ecto query is
```
#Ecto.Query<from s0 in "schema_migrations", select: type(s0.version, :integer)>
```
Unfortunately, I am unable to upload the sqlite3 file I am using due to GitHub restrictions.

Some things to note:
- This works fine on 64 bit machines
- Both esqlite and exqlite are using sqlite v3.38 (as in the latest of both repositories)
- I have not changed the schema_migrations schema. It is this one that is getting used: https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/migration/schema_migration.ex#L10

**Software Versions:**
- Elixir 1.12.3-otp-24
- Erlang 24.1.3
- Linux Kernel 5.4.66
- exqlite 0.10.2
- sqlite 3.38
- ecto_sql - 3.380-dev
- ecto 3.7.0
- ecto_sqlite3 - 0.7.4

**Comment by warmwaffles** [comment]
I'll take a look at this soon. I don't have easy access to 32 bit arm at the moment.

Could you do me a favor and blow away your `_build` and then do `V=1 mix deps.compile exqlite`

**Comment by warmwaffles** [comment]
When I get some time this evening I'll try to recreate this issue. What would be really helpful is setting up a dummy project for me to check out and poke around in.

Can you also take a look and downgrade exqlite to `0.9.3`? In the change from [0.9.3 to 0.10.2](https://github.com/elixir-sqlite/exqlite/compare/v0.9.3...HEAD) I added a custom allocator and I am wondering if that is causing any issues. I doubt it but I just want to check all of my bases.

**Comment by alanj853** [comment]
Hi @warmwaffles , here is the dummy project you can use: https://github.com/alanj853/exqlite_testing

I haven't got around to trying the downgrade yet, will hopefully get time tomorrow.

Thanks again.

**Comment by warmwaffles** [comment]
Thank you.

**Comment by warmwaffles** [comment]
@alanj853 forgive me for being a bit of a scrub at this, but what is the toolchain setup you used?

```
export MIX_TARGET=rpi3
mix deps.get
```

**Comment by alanj853** [comment]
@warmwaffles , we are using a custom armv7 toolchain built from the poky distribution (https://www.yoctoproject.org/software-item/poky/). Unfortunately, I am unable to share the details of it as it is proprietary.

But I do know it is very similar to the toolchain used for cross compiling on the RPI, as I have an RPI and was able to move and run code between the two without issue.

**Comment by alanj853** [comment]
Just another update before anyone else tries it, this only seems to happen on 32 arm. On 32 bit debian/ubuntu for example the issue does not show, even though if you were to do a C program, you would still see the overflow, as expected:

```
user@debian:~/src$ cat main.c 

#include <stdio.h>

int main()

{

	long int x = 20180801160101;

	printf("This is x: %ld\n", x);

	return 0;

}

user@debian:~/src$ gcc -o app main.c 

main.c: In function ‘main’:

main.c:4:15: warning: overflow in conversion from ‘long long int’ to ‘long int’ changes value from ‘20180801160101’ to ‘-1250163803’ [-Woverflow]

    4 |  long int x = 20180801160101;

      |               ^~~~~~~~~~~~~~

user@debian:~/src$ ./app 

This is x: -1250163803

user@debian:~/src$ 
```

**Comment by LostKobrakai** [comment]
I just tested this on one of the devices we work on:
```
iex(31)> from s in "schema_migrations", select: type(s.version, :integer)
#Ecto.Query<from s0 in "schema_migrations", select: type(s0.version, :integer)>
iex(32)> Edith.Repo.all(v())
[20191126141519, 20201103014236, 20210318115442, 20210325093504, 20210325154619,
 20210330165324, 20210331150226, 20210331152036, 20210409080323, 20210423144357,
 20210503132201, 20210505132539, 20210510112318, 20210521151734, 20210527121555,
 20210528100135, 20210601084357, 20210609155612, 20210616161950, 20210618171751,
 20210623152605, 20210705105118, 20210712075610, 20210712165900, 20210716093708,
 20210716151710, 20210721155244, 20210721184024, 20210729125641, 20211109165743,
 20211110125343, 20211117164706, 20211119155759, 20211130140525, 20211202164113,
 20211207160410, 20220121142908, 20220210111737, 20220215123618]
iex(33)> cmd "cat /proc/cpuinfo"
[…]
model name      : ARMv7 Processor rev 2 (v7l)
[…]
```
This is built using a custom nerves system based on the official bbb one, with the nerves toolchain.

**Comment by alanj853** [comment]
I got my hands on an actual RPI3 this morning and tested it out, compiling directly on the PI. No issue. Looking like it might be specific to my toolchain/board setup, not even 32 bit ARM.

```
pi@raspberrypi:~/src/exqlite_testing $ V=1 iex -S mix
Erlang/OTP 24 [erts-12.1.3] [source] [smp:4:4] [ds:4:4:10] [async-threads:1]

==> exqlite
mkdir -p /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/priv
mkdir -p /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/obj
 CC sqlite3_nif.o
cc -c -Ic_src -DNDEBUG=1 -I"/home/pi/.asdf/installs/erlang/24.1.3/erts-12.1.3/include" -fPIC -fvisibility=hidden -DSQLITE_USE_URI -DSQLITE_ENABLE_COLUMN_METADATA -DSQLITE_THREADSAFE=1 -DSQLITE_DEFAULT_WAL_SYNCHRONOUS=3 -DSQLITE_OMIT_DEPRECATED=1 -DSQLITE_OMIT_SHARED_CACHE=1 -DSQLITE_ENABLE_COLUMN_METADATA=1 -DSQLITE_DISABLE_FTS3 -DSQLITE_ENABLE_LOCKING_STYLE=POSIX -DSQLITE_LIKE_DOESNT_MATCH_BLOBS -DSQLITE_ENABLE_UNLOCK_NOTIFY=1 -DSQLITE_USE_URI -DSQLITE_TEMP_STORE=2 -DUSE_PREAD -o /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/obj/sqlite3_nif.o c_src/sqlite3_nif.c
 CC sqlite3.o
cc -c -Ic_src -DNDEBUG=1 -I"/home/pi/.asdf/installs/erlang/24.1.3/erts-12.1.3/include" -fPIC -fvisibility=hidden -DSQLITE_USE_URI -DSQLITE_ENABLE_COLUMN_METADATA -DSQLITE_THREADSAFE=1 -DSQLITE_DEFAULT_WAL_SYNCHRONOUS=3 -DSQLITE_OMIT_DEPRECATED=1 -DSQLITE_OMIT_SHARED_CACHE=1 -DSQLITE_ENABLE_COLUMN_METADATA=1 -DSQLITE_DISABLE_FTS3 -DSQLITE_ENABLE_LOCKING_STYLE=POSIX -DSQLITE_LIKE_DOESNT_MATCH_BLOBS -DSQLITE_ENABLE_UNLOCK_NOTIFY=1 -DSQLITE_USE_URI -DSQLITE_TEMP_STORE=2 -DUSE_PREAD -o /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/obj/sqlite3.o c_src/sqlite3.c
 LD sqlite3_nif.so
cc -o /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/priv/sqlite3_nif.so /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/obj/sqlite3_nif.o /home/pi/src/exqlite_testing/_build/dev/lib/exqlite/obj/sqlite3.o -fPIC -shared
Compiling 10 files (.ex)
Generated exqlite app
Interactive Elixir (1.12.1) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> ExqliteTesting.show_problem
[20220329202356, 20220330122205]
iex(2)>
```

**Comment by warmwaffles** [comment]
@alanj853 keep me posted on any findings please.

**Comment by alanj853** [comment]
I figured out my problem. The toolchain I am using has a different definition for the size of long integers in erl_int_sizes_config.h than the standard erlang one. However, the Makefile wasn't listing my toolchain's erl_int_sizes_config.h directory in the include flags.

Having gone through the issues for this repository, I saw @warmwaffles you mentioned you took some of your Makefile from circuits_gpio. I use this repo as well, so I had a look in that repo's Makefile to see how what they do for cross compiling, and so I have implemented something very similar [here](https://github.com/se-apc/exqlite/pull/2).

I have created a fork on this repo, but let me know @warmwaffles if you want me to merge into this repo as well.
